sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/BufferHolder.java

```diff
@@ -55,6 +58,15 @@ public BufferHolder(UnsafeRow row, int initialSize) {
     this.buffer = new byte[fixedSize + initialSize];
     this.row = row;
     this.row.pointTo(buffer, buffer.length);
+    this.array = null;
+  }
+
+  public BufferHolder(UnsafeArrayData array, long numElements) {
```

Comment:
I cannot understand what this means.
Current `UnsafeArrayWriter` assumes existence of `BufferHolder`. Do you want not to allocate `BufferHolder` for this case and refactor `UnsafeArrayWriter` to support a case without `BufferHolder`?

Or, do you want to remove `numElements`?

Or, do you want to allocate buffer here? This approach introduces duplication of the same logic here and in `UnsafeArrayWriter`.